aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/auth/signing/AwsSignerNative.kt

```diff
@@ -14,30 +27,234 @@ public actual object AwsSigner {
+    val signature = runBlocking { callbackChannel.receive() }.also {
+        callbackChannelStableRef.dispose()
+        callbackChannel.close()
+    }
+
+    return AwsSigningResult(null, signature)
+}
+
+/**
+ * Get the signature of a given [aws_signing_result].
+ * The signature is found in the "signature" property on the signing result.
+ */
+private fun CPointer<aws_signing_result>.getSignature(): ByteArray {
+    val signature = Allocator.Default.allocPointerTo<aws_string>()
+    val propertyName = "signature".toAwsString()
+
+    awsAssertOpSuccess(aws_signing_result_get_property(this, propertyName, signature.ptr)) {
+        "aws_signing_result_get_property"
+    }
+
+    val kSignature = signature.value?.toKString()?.encodeToByteArray()
+    return checkNotNull(kSignature) { "signature was null" }
+}
+
+private fun AwsSigningAlgorithm.toNativeSigningAlgorithm(): aws_signing_algorithm = when (this) {
+    AwsSigningAlgorithm.SIGV4 -> aws_signing_algorithm.AWS_SIGNING_ALGORITHM_V4
+    AwsSigningAlgorithm.SIGV4_ASYMMETRIC -> aws_signing_algorithm.AWS_SIGNING_ALGORITHM_V4_ASYMMETRIC
+}
+
+private fun AwsSignatureType.toNativeSignatureType(): aws_signature_type = when (this) {
+    AwsSignatureType.HTTP_REQUEST_VIA_HEADERS -> aws_signature_type.AWS_ST_HTTP_REQUEST_HEADERS
+    AwsSignatureType.HTTP_REQUEST_VIA_QUERY_PARAMS -> aws_signature_type.AWS_ST_HTTP_REQUEST_QUERY_PARAMS
+    AwsSignatureType.HTTP_REQUEST_EVENT -> aws_signature_type.AWS_ST_HTTP_REQUEST_EVENT
+    AwsSignatureType.HTTP_REQUEST_CHUNK -> aws_signature_type.AWS_ST_HTTP_REQUEST_CHUNK
+    AwsSignatureType.HTTP_REQUEST_TRAILING_HEADERS -> aws_signature_type.AWS_ST_HTTP_REQUEST_TRAILING_HEADERS
+}
+
+private fun AwsSignedBodyHeaderType.toNativeSignedBodyHeaderType() = when (this) {
+    AwsSignedBodyHeaderType.NONE -> aws_signed_body_header_type.AWS_SBHT_NONE
+    AwsSignedBodyHeaderType.X_AMZ_CONTENT_SHA256 -> aws_signed_body_header_type.AWS_SBHT_X_AMZ_CONTENT_SHA256
+}
+
+private fun AwsSigningConfig.toNativeSigningConfig(): CPointer<aws_signing_config_aws> {
+    val config = Allocator.Default.alloc<aws_signing_config_aws>()
+
+    config.apply {
+        config_type = AWS_SIGNING_CONFIG_AWS
+        algorithm = this@toNativeSigningConfig.algorithm.toNativeSigningAlgorithm()
+        signature_type = this@toNativeSigningConfig.signatureType.toNativeSignatureType()
+        region.initFromCursor(this@toNativeSigningConfig.region.toAwsString().asAwsByteCursor())
+        service.initFromCursor(this@toNativeSigningConfig.service.toAwsString().asAwsByteCursor())
+        aws_date_time_init_epoch_millis(date.ptr, this@toNativeSigningConfig.date.toULong())
+
+        // FIXME Can't convert Kotlin config's shouldSignHeader to a C function without capturing the config variable, and staticCFunction cannot capture variables.
+        // should_sign_header = this@toNativeSigningConfig.shouldSignHeader?.toNativeShouldSignHeaderFn()
+        // should_sign_header_ud =
```

Comment:
**Question**: Is there no way to avoid capturing references here? The `shouldSignHeader` param is just a `(String) -> Boolean` so I'm surprised it's causing an issue.

Reply:
I tried a second attempt, but `staticCFunction` can't capture any state, not even the `shouldSignHeader` param. The native should_sign_header function can only use global-level declarations.

https://kotlinlang.org/docs/mapping-function-pointers-from-c.html#pass-kotlin-function-as-c-function-pointer

Do you have any ideas I might be missing here?